tests/Basics/SupervisedBufferTest.cpp

```diff
@@ -0,0 +1,479 @@
+#include "gtest/gtest.h"
+
+#include "Aql/AqlValue.h"
+#include "Basics/Exceptions.h"
+#include "Basics/GlobalResourceMonitor.h"
+#include "Basics/ResourceUsage.h"
+#include "Basics/SupervisedBuffer.h"
+#include <velocypack/Builder.h>
+#include <string>
+
+using namespace arangodb;
+using namespace arangodb::aql;
+using namespace arangodb::velocypack;
+
+TEST(SupervisedbufferTest, AccountsMemoryLargeAndSmallValuesNormalBuffer) {
+  auto& global = GlobalResourceMonitor::instance();
+  ResourceMonitor monitor{global};
+
+  {
+    ResourceUsageScope usageScope(monitor);
+    AqlValue largeValue;
+    {
+      Builder builder;
+      builder.openArray();
+      builder.add(Value(std::string(1024, 'a')));
+      builder.close();
+
+      ASSERT_EQ(monitor.current(), 0);
+      largeValue = AqlValue{builder.slice(), builder.size()};
+      monitor.increaseMemoryUsage(largeValue.memoryUsage());
+
+      ASSERT_EQ(monitor.current(), largeValue.memoryUsage());
+    }
+    ASSERT_EQ(monitor.current(), largeValue.memoryUsage());
+
+    monitor.decreaseMemoryUsage(largeValue.memoryUsage());
+    largeValue.destroy();
+    ASSERT_EQ(monitor.current(), 0);
+  }
+
+  {
+    ResourceUsageScope usageScope(monitor);
+    AqlValue smallValue;
+    {
+      Builder builder;
+      builder.openArray();
+      builder.add(Value(1));
+      builder.close();
+      ASSERT_EQ(monitor.current(), 0);
+      smallValue = AqlValue{builder.slice(), builder.size()};
+      // memoryUsage for small values  is zero
+      // because they fit in the AqlValue's inline buffer
+      monitor.increaseMemoryUsage(smallValue.memoryUsage());
+      ASSERT_EQ(monitor.current(), smallValue.memoryUsage());
+    }
+    // is the same as comparing to smallValue.memoryUsage
+    ASSERT_EQ(monitor.current(), 0);
+
+    monitor.decreaseMemoryUsage(smallValue.memoryUsage());
+    smallValue.destroy();
+    ASSERT_EQ(monitor.current(), 0);
+  }
+}
+
+TEST(SupervisedbufferTest, AccountsMemoryLargeAndSmallValuesSupervisedBuffer) {
+  auto& global = GlobalResourceMonitor::instance();
+  ResourceMonitor monitor{global};
+
+  {
+    ResourceUsageScope usageScope(monitor);
+    AqlValue largeValue;
+    {
+      SupervisedBuffer supervisedBuffer(monitor);
+      Builder builder(supervisedBuffer);
+      builder.openArray();
+      builder.add(Value(std::string(1024, 'a')));
+      builder.close();
+
+      ASSERT_GT(monitor.current(), 0);
```

Comment:
can we get this a bit more precise?
A string of 1024 'a's has at least 1024 byte of memory, not at least 0.

same later on the GE assertion.